### PR TITLE
Add ops the ability to clean orphan commands

### LINKED ISSUE #1397: Mark orphan commands with TIMEOUT status
**What would you like to be added**:

Currently, piped is using the following gRPC APIs to fetch the unhandled commands and save the status of handled ones.
https://github.com/pipe-cd/pipe/blob/master/pkg/app/api/grpcapi/piped_api.go#L544-L601

But for some reason, a command will not be handled or its status was not saved correctly to the control-plane.
We call them orphan commands. (e.g. commands created more than 1 week ago but still not be handled)

And we have to periodically find them and mark them with `TIMEOUT` status:
https://github.com/pipe-cd/pipe/blob/master/pkg/app/piped/apistore/commandstore/store.go
https://github.com/pipe-cd/pipe/blob/master/pkg/model/command.proto#L27

**Why is this needed**:

### PR DESCRIPTION
**Pull request by sanposhiho**
**What this PR does / why we need it**:

Create orphan commands cleaner. It periodically find orphan commands and mark them with TIMEOUT status

**Which issue(s) this PR fixes**:

Fixes #1397 

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/app/ops/orphancommandcleaner/orphancommandcleaner.go [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+	"go.uber.org/zap"
+
+	"github.com/pipe-cd/pipe/pkg/datastore"
+	"github.com/pipe-cd/pipe/pkg/model"
+)
+
+var (
+	commandTimeOut = 7 * 24 * time.Hour
+	interval       = 24 * time.Hour
```
Comment (sanposhiho) [inline pkg/app/ops/orphancommandcleaner/orphancommandcleaner.go:15]: What values should be set?

**Comment by sanposhiho** [comment]
Is there a better place to create `orphan commands cleaner` instead of `ops`?

File: cmd/pipecd/ops.go [hunk 2]
```diff
@@ -98,6 +99,10 @@ func (s *ops) run(ctx context.Context, t cli.Telemetry) error {
 		}
 	}()
 
+	// Starting orphan commands cleaner
+	cleaner := orphancommandcleaner.NewOrphanCommandCleaner(ds, t.Logger)
+	cleaner.Run(ctx)
```
Comment (nghialv) [inline cmd/pipecd/ops.go:104]: This should be wrapped by `errgroup` to terminate `ops` when `cleaner` exit because of an error.

``` golang
group.Go(func() error {
	return cleaner.Run(ctx)
})
```

File: pkg/app/ops/orphancommandcleaner/orphancommandcleaner.go (hunk 1 above)
Comment (nghialv) [reply pkg/app/ops/orphancommandcleaner/orphancommandcleaner.go:15]: `commandTimeOut = 24 * time.Hour`
` interval = 6 * time.Hour`

File: pkg/app/ops/orphancommandcleaner/orphancommandcleaner.go [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+	logger       *zap.Logger
+}
+
+func NewOrphanCommandCleaner(
+	ds datastore.DataStore,
+	logger *zap.Logger,
+) *OrphanCommandCleaner {
+	return &OrphanCommandCleaner{
+		commandstore: datastore.NewCommandStore(ds),
+		logger:       logger.Named("orphan-command-finder"),
```
Comment (nghialv) [inline pkg/app/ops/orphancommandcleaner/orphancommandcleaner.go:29]: `orphan-command-cleaner`

File: pkg/app/ops/orphancommandcleaner/orphancommandcleaner.go [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+	}
+}
+
+func (o *OrphanCommandCleaner) Run(ctx context.Context) {
+	for {
+		if err := o.updateOrphanCommandsStatus(ctx); err != nil {
+			o.logger.Error("failed to update orphan commands", zap.Error(err))
+		}
+		time.Sleep(interval)
```
Comment (nghialv) [inline pkg/app/ops/orphancommandcleaner/orphancommandcleaner.go:38]: Let's use `time.Timer` instead of sleep.

File: pkg/app/ops/orphancommandcleaner/orphancommandcleaner.go [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+	ds datastore.DataStore,
+	logger *zap.Logger,
+) *OrphanCommandCleaner {
+	return &OrphanCommandCleaner{
+		commandstore: datastore.NewCommandStore(ds),
+		logger:       logger.Named("orphan-command-finder"),
+	}
+}
+
+func (o *OrphanCommandCleaner) Run(ctx context.Context) {
```
Comment (nghialv) [inline pkg/app/ops/orphancommandcleaner/orphancommandcleaner.go:33]: nit: `func(c *` because it is a cleaner.

File: pkg/app/ops/orphancommandcleaner/orphancommandcleaner.go [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+	for {
+		if err := o.updateOrphanCommandsStatus(ctx); err != nil {
+			o.logger.Error("failed to update orphan commands", zap.Error(err))
+		}
+		time.Sleep(interval)
+	}
+}
+
+func (o *OrphanCommandCleaner) updateOrphanCommandsStatus(ctx context.Context) error {
+	timeoutedTime := time.Now().Add(-commandTimeOut).Unix()
```
Comment (nghialv) [inline pkg/app/ops/orphancommandcleaner/orphancommandcleaner.go:43]: nit: `timedout` or just `timeout`

File: pkg/app/ops/orphancommandcleaner/orphancommandcleaner.go [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+			},
+		},
+	}
+	commands, err := o.commandstore.ListCommands(ctx, opts)
+	if err != nil {
+		return err
+	}
+
+	for _, c := range commands {
+		o.commandstore.UpdateCommand(ctx, c.Id, func(c *model.Command) error {
```
Comment (nghialv) [inline pkg/app/ops/orphancommandcleaner/orphancommandcleaner.go:64]: Can you add a log to show that the command was successfully updated or not?

**Comment by nghialv** [comment]
> Is there a better place to create orphan commands cleaner instead of ops?

I think the current place is fine.

**Comment by nghialv** [comment]
@sanposhiho Hi, Thank you for your great effort.
I have just left some comments. Please take a look.

File: cmd/pipecd/ops.go (hunk 2 above)
Comment (sanposhiho) [reply cmd/pipecd/ops.go:104]: IMO, there was no need to stop ops because errors that occur in cleaner is probably caused by a network error when communicating with datastore.

Of course, it is necessary to output the error log to show that an error has occurred. However, I don't think it is necessary to stop ops because updateOrphanCommandsStatus will include the update that should have been done with the previous process if the previous process failed.

Or how about stopping `ops` if it fails 3 times in a row?

What do you think? 🤔

Comment (nghialv) [reply cmd/pipecd/ops.go]: Nice comment. I got your points.
Agree with you that the `cleaner` may have some errors while executing its jobs.

But that is the problem of `cleaner`, not this `run` function.
This `run` should not care about what each component is doing, it just cares that they are fine or not. If any of them had a problem, `ops` should be stopped to notify that there was a problem from one of its components.
In other words, `run` should see all of its components (`cleaner`, `http-server`, `insightcollector`...) in the same way. 

About the network problem or something that can cause the error while `cleaner` executing the job,
`cleaner` can ignore them,  just log the error message, and does not return. 
By that way, the `ops` will not be stopped by those errors and we can keep `run` to be nice too.

File: pkg/app/ops/orphancommandcleaner/orphancommandcleaner.go [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+		logger:       logger.Named("orphan-command-cleaner"),
+	}
+}
+
+func (c *OrphanCommandCleaner) Run(ctx context.Context) error {
+	consecutiveFailuresCount := 0
+
+	t := time.NewTicker(interval)
+	for {
+		select {
```
Comment (nghialv) [inline pkg/app/ops/orphancommandcleaner/orphancommandcleaner.go:39]: Please check the `context.Done` to exit when `ops` was terminated.

``` golang
for {
    select {
          case <- ctx.Done():
                 return nil
         case <- t.C:
             ...
}
}
```

File: pkg/app/ops/orphancommandcleaner/orphancommandcleaner.go [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+	if err != nil {
+		return err
+	}
+
+	for _, command := range commands {
+		if err := c.commandstore.UpdateCommand(ctx, command.Id, func(cmd *model.Command) error {
+			cmd.Status = model.CommandStatus_COMMAND_TIMEOUT
+			return nil
+		}); err != nil {
+			c.logger.Error("failed to update orphan commands", zap.Error(err))
```
Comment (nghialv) [inline pkg/app/ops/orphancommandcleaner/orphancommandcleaner.go:80]: nit: `commands` - > `command`
And can you add the command ID and Type to the log message?

File: pkg/app/ops/orphancommandcleaner/orphancommandcleaner.go [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+
+	t := time.NewTicker(interval)
+	for {
+		select {
+		case <-t.C:
+			if err := c.updateOrphanCommandsStatus(ctx); err != nil {
+				c.logger.Error("failed to update orphan commands", zap.Error(err))
+				consecutiveFailuresCount++
+				if consecutiveFailuresCount == maxConsecutiveFailuresCount {
+					return err
```
Comment (nghialv) [inline pkg/app/ops/orphancommandcleaner/orphancommandcleaner.go:45]: As I explained [here](https://github.com/pipe-cd/pipe/pull/1434#discussion_r558296643), we can ignore this error.
Don't have to return. And we don't need the error log too because it is already logged inside the `updateOrphanCommandsStatus` function.

Btw, It would be nice if you can add a log to show the `elapsed` time of this execution.

File: cmd/pipecd/ops.go (hunk 2 above)
Comment (sanposhiho) [reply cmd/pipecd/ops.go]: > About the network problem or something that can cause the error while cleaner executing the job,
cleaner can ignore them, just log the error message, and does not return.
By that way, the ops will not be stopped by those errors and we can keep run to be nice too.

OK, I see 👍

File: pkg/app/ops/orphancommandcleaner/orphancommandcleaner.go (hunk 8 above)
Comment (sanposhiho) [reply pkg/app/ops/orphancommandcleaner/orphancommandcleaner.go:39]: Oops, thanks

**Comment by sanposhiho** [comment]
@nghialv 
Thanks for your reviews, fixed them :D

**Comment by nghialv** [comment]
@sanposhiho Thank you very much for your great work! ❤️ 
/approve
